src/wrapper.ts

```diff
@@ -6,158 +6,131 @@ import * as api from './api';
 import * as legacy from './legacy';
 import * as client from './client';
 import * as util from './util';
-import {config} from './config';
+import { config } from './config';
+import { log } from './logging';
 
-export class CMakeToolsWrapper implements api.CMakeToolsAPI {
-  private _impl: Promise<api.CMakeToolsAPI>;
+export class CMakeToolsWrapper {
+  private _impl?: api.CMakeToolsAPI = undefined;
 
-  constructor(private _ctx: vscode.ExtensionContext) {}
+  constructor(private _ctx: vscode.ExtensionContext) { }
 
-  async dispose() {
+  public async dispose() {
     await this.shutdown();
     this._reconfiguredEmitter.dispose();
   }
 
-  private async _sourceDir() {
-    return (await this._impl).sourceDir;
-  }
-  get sourceDir() {
-    return this._sourceDir();
+  public get toolsApi(): api.CMakeToolsAPI {
+    if (this._impl)
+      return this._impl;
+    else
+      throw new Error("CMakeTools is not initialized");
   }
 
-  private async _mainListFile() {
-    return (await this._impl).mainListFile;
-  }
-  get mainListFile() {
-    return this._mainListFile();
-  }
-
-  private async _binaryDir() {
-    return (await this._impl).binaryDir;
-  }
-  get binaryDir() {
-    return this._binaryDir();
-  }
-
-  private async _cachePath() {
-    return (await this._impl).cachePath;
-  }
-  get cachePath() {
-    return this._cachePath();
-  }
-
-  private async _executableTargets() {
-    return (await this._impl).executableTargets;
-  }
-  get executableTargets() {
-    return this._executableTargets();
-  }
-
-  private async _diagnostics() {
-    return (await this._impl).diagnostics;
-  }
-  get diagnostics() {
-    return this._diagnostics();
-  }
-
-  private async _targets() {
-    return (await this._impl).targets;
-  }
-  get targets() {
-    return this._targets();
+  public async configure(extraArgs?: string[], runPrebuild?: boolean) {
+    if (this._impl)
```

Comment:
Alternatively this.toolsApi may be used here with some try/catch logic in top-level command execution.
E.g. configure() has it because cmake server may fail to (re)start.